### PR TITLE
Improve tuned test suite

### PR DESCRIPTION
**Pull request by marioferh**
## Description
Improve legibility and facility to add test in tuned test suite. 
Remove Regex and parse data from template

## Motivation and Context
Difficulty to add new test and understand the current ones

## How has this been tested?
Test are launched in ci under ci/prow/e2e-no-cluster  lane

## Types of changes

- [x] Parse INI structure of tuned template to improve test legibility
- [x] Delete regex in order to reduce complexity 
- [x] Re order test. Improve different combinations at the expense of little more verbosity. Add all possible combinations 

Signed-off-by: Mario Fernandez <mariofer@redhat.com>

### CODE REVIEW

**Review by bartwensley — CHANGES_REQUESTED** [review]

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 1]
```diff
@@ -51,87 +52,148 @@ var _ = Describe("Tuned", func() {
...
 				profile.Spec.WorkloadHints = &performancev2.WorkloadHints{RealTime: pointer.BoolPtr(false)}
+				tunedData := getTunedStructuredData(profile)
+				bootLoader, err := tunedData.GetSection("bootloader")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(bootLoader.Key("cmdline_realtime").String()).ShouldNot(Equal(cmdlineRealtime))
 			})
+		})
 
+		When("realtime hint enabled", func() {
 			It("should not contain realtime related parameters", func() {
```
[the remark this answers is not in the document]
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:103]: done

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 2]
```diff
@@ -51,87 +52,148 @@ var _ = Describe("Tuned", func() {
...
 		})
 
 		When("high power consumption hint enabled", func() {
-			BeforeEach(func() {
+			It("should contain high power consumption related parameters", func() {
 				profile.Spec.WorkloadHints = &performancev2.WorkloadHints{HighPowerConsumption: pointer.BoolPtr(true)}
+				tunedData := getTunedStructuredData(profile)
+				bootLoader, err := tunedData.GetSection("bootloader")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(bootLoader.Key("cmdline_power_performance").String()).ShouldNot(Equal(cmdlineHighPowerConsumption))
```
[the remark this answers is not in the document]
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:128]: right, its should, it was an extra space at the end of the string that do the test works... Fix it

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 3]
```diff
@@ -51,87 +52,148 @@ var _ = Describe("Tuned", func() {
...
 			})
+		})
 
-			It("should contain high power consumption related parameters", func() {
-				manifest := getTunedManifest(profile)
-				By("Populating high power consumption cmdline")
-				Expect(cmdlineHighPowerConsumption.MatchString(manifest)).To(BeTrue())
-				By("Populating idle poll cmdline")
-				Expect(cmdlineHighPowerConsumptionWithRealtime.MatchString(manifest)).To(BeTrue())
+		When("high power consumtptoin enabled + realtime workload hint enabled", func() {
```
[the remark this answers is not in the document]
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:145]: done

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 4]
```diff
@@ -51,87 +52,148 @@ var _ = Describe("Tuned", func() {
...
+			It("should not contain idle=poll cmdline", func() {
```
[the remark this answers is not in the document]
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:146]: done

**Review by bartwensley — APPROVED** [review]
Looks good. BTW - it would be much easier to review if you would push new commits to address the CI comments - that way I would be able to see what you changed and mark the comments as resolved. When you force push, I have to go back and look at the whole thing and manually figure out what you changed.

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 5]
```diff
@@ -22,20 +23,18 @@ const expectedMatchSelector = `
...
-	cmdlineMultipleHugePages                = regexp.MustCompile(`\s*cmdline_hugepages=\+\s*default_hugepagesz=1G\s+hugepagesz=1G\s+hugepages=4\s+hugepagesz=2M\s+hugepages=128\s*`)
+	cmdlineCPUsPartitioning       = "+nohz=on rcu_nocbs=${isolated_cores} tuned.non_isolcpus=${not_isolated_cpumask} systemd.cpu_affinity=${not_isolated_cores_expanded} intel_iommu=on iommu=pt"
+	cmdlineWithStaticIsolation    = "+isolcpus=domain,managed_irq,${isolated_cores}"
+	cmdlineWithoutStaticIsolation = "+isolcpus=managed_irq,${isolated_cores}"
+	cmdlineRealtime               = "+nohz_full=${isolated_cores} tsc=nowatchdog nosoftlockup nmi_watchdog=0 mce=off skew_tick=1 rcutree.kthread_prio=11"
+	cmdlineHighPowerConsumption   = "+processor.max_cstate=1 intel_idle.max_cstate=0 intel_pstate=disable"
+	cmdlineIdlePoll               = "+idle=poll"
+	cmdlineHugepages              = "+ default_hugepagesz=1G   hugepagesz=1G hugepages=4"
+	cmdlineAdditionalArgs         = "+ audit=0 processor.max_cstate=1 idle=poll intel_idle.max_cstate=0"
```
Comment (swatisehgal) [inline pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:33]: I see we have moved away from `"test1=val1", "test2=val2"` to more realistic examples for kernel params.  IIUC, we are providing a set of  random kernel arguments here and later checking against them. Is my understanding here correct? If that is the case, lets add a comment here that this is an example of how additional kernel arguments could look like and these arguments have been selected at random for demonstration purposes and have no other significance.

**Review by swatisehgal — COMMENTED** [review]
Thanks for the PR. Apologies for not being able to get to it sooner. 

The PR is looking good. I left a few comments with possible improvements and clarification questions for better understanding.

Also, would have been nicer if we had multiple commits and the commit breakdown aligned with the "Types of changes" section in the PR description but this is not a blocker in any way.

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 6]
```diff
@@ -51,87 +50,148 @@ var _ = Describe("Tuned", func() {
...
+				tunedData := getTunedStructuredData(profile)
+				service, err := tunedData.GetSection("service")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(service.Key("service.stalld").String()).Should(Equal("start,enable"))
+				scheduler, err := tunedData.GetSection("scheduler")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(scheduler.Key("sched_rt_runtime_us").String()).Should(Equal("-1"))
+				sysctl, err := tunedData.GetSection("sysctl")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(sysctl.Key("kernel.hung_task_timeout_secs").String()).Should(Equal("600"))
```
Comment (Tal-or) [inline pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:112]: According to: https://github.com/onsi/gomega/blob/master/gomega_dsl.go#L415  ```To```/```ToNot```/```NotTo``` are used with ```Expect```

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 7]
```diff
@@ -51,87 +50,148 @@ var _ = Describe("Tuned", func() {
...
+				Expect(sysctl.Key("kernel.sched_rt_runtime_us").String()).Should(Equal("-1"))
+				bootLoader, err := tunedData.GetSection("bootloader")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(bootLoader.Key("cmdline_realtime").String()).Should(Equal(cmdlineRealtime))
```
[the remark this answers is not in the document]
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:116]: Yep, the idea was split the checks and do it in different When and It
In fact the the title is: 
Context("with worker performance profile", func() {
		It("should generate yaml with expected parameters", 

and inside the test we are checking all parameters and its to general. But yes, maybe its better to check it at both places.

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 8]
```diff
@@ -51,87 +50,148 @@ var _ = Describe("Tuned", func() {
...
-			By("Populating CPU partitioning cmdline")
-			Expect(cmdlineCPUsPartitioning.MatchString(manifest)).To(BeTrue())
-			By("Populating static isolation cmdline")
-			Expect(cmdlineWithStaticIsolation.MatchString(manifest)).To(BeTrue())
-			By("Populating hugepages cmdline")
-			Expect(cmdlineHugepages.MatchString(manifest)).To(BeTrue())
-			By("Populating empty additional kernel arguments cmdline")
-			Expect(manifest).To(ContainSubstring("cmdline_additionalArg="))
-			By("Populating realtime cmdline")
-			Expect(cmdlineWithRealtime.MatchString(manifest)).To(BeTrue())
```
[the remark this answers is not in the document]
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:74]: Answer in next comment.

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 9]
```diff
@@ -51,87 +50,148 @@ var _ = Describe("Tuned", func() {
...
 		})
 
 		When("realtime hint disabled", func() {
-			BeforeEach(func() {
+			It("should not contain realtime related parameters", func() {
 				profile.Spec.WorkloadHints = &performancev2.WorkloadHints{RealTime: pointer.BoolPtr(false)}
+				tunedData := getTunedStructuredData(profile)
+				bootLoader, err := tunedData.GetSection("bootloader")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(bootLoader.Key("cmdline_realtime").String()).ShouldNot(Equal(cmdlineRealtime))
```
[the remark this answers is not in the document]
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:96]: Done

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go [hunk 10]
```diff
@@ -51,87 +50,148 @@ var _ = Describe("Tuned", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+				Expect(sysctl.Key("kernel.hung_task_timeout_secs").String()).Should(Equal("600"))
+				Expect(sysctl.Key("kernel.sched_rt_runtime_us").String()).Should(Equal("-1"))
+				bootLoader, err := tunedData.GetSection("bootloader")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(bootLoader.Key("cmdline_realtime").String()).Should(Equal(cmdlineRealtime))
 			})
 		})
 
 		When("high power consumption hint enabled", func() {
```
[the remark this answers is not in the document]
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:120]: Done. Tidier. Thanks

File: pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go (hunk 5 above)
Comment (marioferh) [reply pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned_test.go:33]: Done, right. I had moved these test1=val1 to the testing.go profile and add a more "real ones". To have everything in the same file and tidier. 
Add comment before the test.

**Comment by swatisehgal** [comment]
/lgtm

Thanks for addressing all the review comments.

**Pull request merged**